File: naturals/naturals.hs
data Nat = Zero | Suc Nat

eq :: Nat -> Nat -> Bool
eq Zero Zero = True
eq Zero (Suc m) = False
eq (Suc n) Zero = False
eq (Suc n) (Suc m) = eq n m

leq :: Nat -> Nat -> Bool
leq Zero _ = True
leq _ Zero = False
leq (Suc n) (Suc m) = leq n m

plusOne :: Nat -> Nat
plusOne n = Suc n

add :: Nat -> Nat -> Nat
add Zero m = m
add n Zero = n
add n (Suc m) = add (Suc n) m

mult :: Nat -> Nat -> Nat
mult Zero _ = Zero
mult _ Zero = Zero
mult n (Suc m) = add n (mult n m) 

showsNat :: Nat -> String -> String
showsNat Zero s = show $ s ++ "zero"
showsNat (Suc n) s = showsNat n (s ++ "suc ")

instance Show Nat where
	showsPrec _ n = showsNat n
	
zero :: Nat
zero = Zero

one :: Nat
one = Suc zero

two :: Nat
two = Suc one

File: naturals/naturals.c
#include <stdio.h>
#include <stdlib.h>
#define zero NULL

typedef struct {
	struct Nat* suc;
} Nat;

void printNat(Nat* nat) {
	while (nat) {
		printf("suc ");
		nat = nat->suc;
	}
	printf("zero\n");
}

short eq(Nat* nat1, Nat* nat2) {
	if (!nat1) return !nat2;
	if (!nat2) return 0;
	return eq(nat1->suc, nat2->suc);
}

short leq(Nat* nat1, Nat* nat2) {
	if (!nat1) return 1;
	if (!nat2) return 0;
	return leq(nat1->suc, nat2->suc);
}

Nat* plusOne(Nat* nat) {
	Nat* res = (Nat*)malloc(sizeof(Nat));
	res->suc = nat;
	return res;
}

Nat* add(Nat* nat1, Nat* nat2) {
	if (!nat2) return nat1;
	Nat* res = (Nat*)malloc(sizeof(Nat));
	res->suc = nat1;

	return add(res, nat2->suc);
}

Nat* mult(Nat* nat1, Nat* nat2) {
	if (!nat1 || !nat2) return zero;
	if (nat2->suc == zero) return nat1;

	Nat* res = zero;
	for (Nat* iter = nat2; iter != zero; iter = iter->suc) {
		res = add(nat1, res);
	}

	return res;
}


int main() {
	Nat* one = (Nat*)malloc(sizeof(Nat));
	Nat* two = (Nat*)malloc(sizeof(Nat));
	one->suc = zero;
	two->suc = one;

	printNat(zero);
	printNat(one);
	printNat(two);

    printf("1 = 1 ? %d \n", eq(one, one));
    printf("1 = 1 ? %d \n", eq(two, one));
    printf("1 = 1 ? %d \n", leq(zero, one));
    printf("1 = 1 ? %d \n", leq(two, one));
    
    printf("1 + 1 = ");
    printNat(plusOne(one));
    printf("2 + 2 = ");
    printNat(add(two, two));
	printf("2 * (2 + 1) = ");
    printNat(mult(two, plusOne(two)));

	return 0;
}